File: frontend/navi-chat/functions/mi-test.ts
/**
 * mi-test.ts — Azure Function (HTTP trigger)
 *
 * Valida que ManagedIdentityCredential pueda obtener un token
 * con scope https://api.fabric.microsoft.com/.default.
 *
 * Despliegue: func azure functionapp publish <app-name>
 * Requiere: npm install @azure/identity
 */

import { AzureFunction, Context, HttpRequest } from "@azure/functions";
import { ManagedIdentityCredential, DefaultAzureCredential } from "@azure/identity";

const FABRIC_SCOPE = "https://api.fabric.microsoft.com/.default";

interface TestResult {
  success: boolean;
  credential: string;
  tokenLength?: number;
  error?: string;
  stack?: string;
}

async function tryCredential(cred, label: string): Promise<TestResult> {
  try {
    const token = await cred.getToken(FABRIC_SCOPE);
    return { success: true, credential: label, tokenLength: token.token.length };
  } catch (err: any) {
    return { success: false, credential: label, error: err.message ?? String(err), stack: err.stack };
  }
}

const httpTrigger: AzureFunction = async function (
  context: Context,
  req: HttpRequest
): Promise<void> {
  const results: TestResult[] = await Promise.all([
    tryCredential(new ManagedIdentityCredential(), "ManagedIdentityCredential"),
    tryCredential(new DefaultAzureCredential(), "DefaultAzureCredential"),
  ]);

  const anyOk = results.some((r) => r.success);
  context.res = {
    status: anyOk ? 200 : 503,
    body: { results },
  };
};

export default httpTrigger;

File: frontend/navi-chat/functions/mi-test.py
"""
mi-test.py
----------
Test mínimo de ManagedIdentityCredential para scope Fabric.

Uso (local con az login):  python functions/mi-test.py
Uso (Azure VM / App Service con MI): python functions/mi-test.py

Requiere: pip install azure-identity python-dotenv
"""

import os
import json
import sys
from pathlib import Path

from azure.identity import DefaultAzureCredential, ManagedIdentityCredential

FABRIC_SCOPE = "https://api.fabric.microsoft.com/.default"


def try_credential(cred, label: str) -> dict:
    try:
        token = cred.get_token(FABRIC_SCOPE)
        return {"success": True, "credential": label, "tokenLength": len(token.token), "expiresOn": token.expires_on}
    except Exception as e:
        return {"success": False, "credential": label, "error": str(e)}


def main():
    results = []

    results.append(try_credential(ManagedIdentityCredential(), "ManagedIdentityCredential"))

    results.append(try_credential(DefaultAzureCredential(), "DefaultAzureCredential"))

    print(json.dumps(results, indent=2))

    any_ok = any(r["success"] for r in results)
    sys.exit(0 if any_ok else 1)


if __name__ == "__main__":
    main()